modules/dnn/src/layers/onehot_layer.cpp

```diff
@@ -0,0 +1,269 @@
+            for (int pos = r.start; pos < r.end; ++pos) {
+                int64_t idxVal = localIdxPtr64[pos];
+                int64_t wrapped = idxVal % localDepth;
+                if (wrapped < 0) wrapped += localDepth;
+                size_t hi = (size_t)pos / localOutStep;
+                size_t lo = (size_t)pos % localOutStep;
+                size_t base = hi * ((size_t)localDepth * localOutStep) + lo;
+                localOutPtr[base + (size_t)wrapped * localOutStep] = localOnVal;
+            }
+        });
+    }
+
+public:
+    void forward(InputArrayOfArrays in, OutputArrayOfArrays out, OutputArrayOfArrays) CV_OVERRIDE
+    {
+        int inKind = in.kind();
+        int outKind = out.kind();
+        CV_Assert(in.size().area() == 3);
+
+        Mat indices, depthMat, values;
+        if (inKind == _InputArray::STD_VECTOR_MAT)
+        {
+            indices = in.getMat(0);
+            depthMat = in.getMat(1);
+            values = in.getMat(2);
+        }
+        else if (inKind == _InputArray::STD_VECTOR_UMAT)
+        {
+            indices = in.getUMat(0).getMat(ACCESS_READ);
+            depthMat = in.getUMat(1).getMat(ACCESS_READ);
+            values = in.getUMat(2).getMat(ACCESS_READ);
+        }
+        else
+        {
+            CV_Error(Error::StsNotImplemented, "OneHot: unsupported input type");
+        }
+
+        std::vector<Mat>* out_mats = nullptr;
+        std::vector<UMat>* out_umats = nullptr;
+        Mat Y;
+        if (outKind == _InputArray::STD_VECTOR_MAT)
+        {
+            out_mats = &out.getMatVecRef();
+            out_mats->resize(1);
+        }
+        else if (outKind == _InputArray::STD_VECTOR_UMAT)
+        {
+            out_umats = &out.getUMatVecRef();
+            out_umats->resize(1);
+        }
+        else
+        {
+            CV_Error(Error::StsNotImplemented, "OneHot: unsupported output type");
+        }
+
+        int outDepth = values.empty() ? CV_32F : values.depth();
+
+        MatShape outShape = shape(indices);
+        int ndims = (int)outShape.size();
+        int insAxis = normalize_axis(axis, ndims + 1);
+
+        CV_Assert(depthMat.total() == 1);
+        int64_t depth64 = 0;
+        tensorToScalar(depthMat, CV_64S, &depth64);
+        CV_CheckGT(depth64, 0, "OneHot: depth must be > 0");
+        MatShape finalShape;
+        outShape.insert(outShape.begin() + insAxis, (int)depth64);
+        finalShape = outShape;
+        if (!finalShape.empty())
+        {
+            if (outKind == _InputArray::STD_VECTOR_MAT)
+            {
+                out_mats->at(0).fit(finalShape, outDepth);
+                Y = out_mats->at(0);
+            }
+            else
+            {
+                out_umats->at(0).fit(finalShape, outDepth);
+                Y = Mat(finalShape, outDepth);
+            }
+        }
+
+        switch (outDepth) {
+            case CV_8U:   runKernel<uchar>(indices, (int)depth64, values, Y); break;
```

Comment:
please, reduce the number of branches. `{8U, 8S and Bool}` is one case, `{16U, 16S, 16F, 16BF}` is another, `{32U, 32S, 32F}` is yet another and finally `{64U, 64S, 64F}`. Always use integers for those cases, don't use float or double, because certain bit patterns may be automatically transformed when you interpret 4 bytes (or 2 bytes or 8 bytes) as floating-point number.